src/cpu/o3/inst_queue.cc

```diff
@@ -525,16 +511,118 @@ InstructionQueue::processFUCompletion(const DynInstPtr &inst, int fu_idx)
    --wbOutstanding;
     iewStage->wakeCPU();
 
-    if (fu_idx > -1)
-        fuPool->freeUnitNextCycle(fu_idx);
-
     // @todo: Ensure that these FU Completions happen at the beginning
     // of a cycle, otherwise they could add too many instructions to
     // the queue.
     issueToExecuteQueue->access(0)->size++;
     instsToExecute.push_back(inst);
 }
 
+bool
+InstructionQueue::execLatencyCheck(const DynInstPtr& inst, uint32_t& op_latency)
+{
+    auto lzc = [](RegVal val) {
+        for (int i = 0; i < 64; i++) {
+            if (val & (0x1lu << 63)) {
+                return i;
+            }
+            val <<= 1;
+        }
+        return 64;
+    };
+    RegVal rs1;
+    RegVal rs2;
+    int delay_;
+    switch (inst->opClass()) {
+        case OpClass::IntDiv:
+            rs1 = cpu->readArchIntReg(inst->srcRegIdx(0).index(),
+                                      inst->threadNumber);
+            rs2 = cpu->readArchIntReg(inst->srcRegIdx(1).index(),
+                                      inst->threadNumber);
+            // rs1 / rs2 : 0x80/0x8 ,delay_ = 4
+            delay_ = std::max(lzc(std::labs(rs2)) - lzc(std::labs(rs1)), 0);
+            if (rs2 == 1) {
+                op_latency = 6;
+            } else if (rs1 == rs2) {
+                op_latency = 8;
+            } else if (lzc(std::labs(rs2)) - lzc(std::labs(rs1)) < 0) {
+                op_latency = 6;
+            } else {
+                op_latency = 8 + delay_ / 4;
+            }
+            return true;
```

Comment:
plz add more comments for division latency computation and sqrt latency computation